Shard RegisterDispatchKey

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144364
* #144363

Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: built PyTorch on Raspberry Pi 5; this seemed to alleviate high peak memory requirement. (I did increase shard counts for other generated files along the way, but I need to go back and figure out how much of that was strictly necessary vs. needing to use -j1 or -j2.)

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

swolchok committed:
Shard RegisterDispatchKey

Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: build PyTorch on Raspberry Pi 5 to demonstrate this fixes high peak memory requirement (TODO).

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

Skylion007 commented on torchgen/gen.py:
```diff
@@ -2305,34 +2305,49 @@ def operator_headers() -> list[str]:
+            env_callable=register_dispatch_key_env_callable,
+            num_shards=4 if dispatch_key == DispatchKey.CPU else 1,
+            base_env=register_dispatch_key_base_env,
+            sharded_keys={"dispatch_definitions"},
```
```suggestion
            sharded_keys=frozenset({"dispatch_definitions"}),
```
Slight nit for immutability of default args

swolchok replied:
there are several other uses of sharded_keys and none of them is a frozenset. how about a mass fixup?

swolchok replied:
#144408

swolchok replied:
I'm confused. This isn't a default argument, it's just a keyword argument at an ordinary callsite.

Skylion007 replied:
Whoops, misread it on my phone. You are right.

swolchok committed:
unbreak gen_backend_stubs

swolchok commented:
whoops, pushed 7df69dbdd04beb59dc96817d22c52bee5a640f4c to the wrong PR (it goes in the bottom-of-stack PR), will fix shortly

swolchok commented:
@swolchok has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67925496).

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: build PyTorch on Raspberry Pi 5 to demonstrate this fixes high peak memory requirement (TODO).

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: build PyTorch on Raspberry Pi 5 to demonstrate this fixes high peak memory requirement (TODO).

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: built PyTorch on Raspberry Pi 5; this seemed to alleviate high peak memory requirement. (I did increase shard counts for other generated files along the way, but I need to go back and figure out how much of that was strictly necessary vs. needing to use -j1 or -j2.)

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

Skylion007 commented on buckbuild.bzl:
```diff
@@ -409,20 +409,22 @@ def get_aten_generated_files(enabled_backends):
-        ":{}[{}]".format(aten_rule_name, "Register" + backend + ".cpp")
-        for backend in enabled_backends
-    ]
+        ":{}[{}]".format(aten_rule_name, "Register" + backend + "_0.cpp")
```
While adjusting this, f-string might be a bit cleaner

swolchok replied:
Starlark doesn't have f-strings -- https://github.com/bazelbuild/starlark/issues/91

bdhirsh commented on torchgen/utils.py:
```diff
@@ -209,6 +209,29 @@ def write_sharded(
+    def write_sharded_with_template(
+        self,
+        filename: str,
+        template_fn: str,
```
nit: `template_fn` -> `template_name`? (it looks like this is the string name of the input template file, not a function)

swolchok replied:
this matches the existing `template_fn` parameter to `write_with_template` and `substitute_with_template`. I agree that it is confusing to overload "fn" to abbreviate both "filename" and "function", and there is a lot of historical precedent for "function", so I will just do a file-wide find-replace.

swolchok replied:
> file-wide find-replace
BC-Lint has informed me that I can't do that, so I will keep it template_fn for consistency :\

bdhirsh reviewed (approved):
sgtm!

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: built PyTorch on Raspberry Pi 5; this seemed to alleviate high peak memory requirement. (I did increase shard counts for other generated files along the way, but I need to go back and figure out how much of that was strictly necessary vs. needing to use -j1 or -j2.)

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: built PyTorch on Raspberry Pi 5; this seemed to alleviate high peak memory requirement. (I did increase shard counts for other generated files along the way, but I need to go back and figure out how much of that was strictly necessary vs. needing to use -j1 or -j2.)

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

swolchok committed:
Update on "Shard RegisterDispatchKey"


Should fix https://github.com/pytorch/pytorch/issues/143952 .

Testing: built PyTorch on Raspberry Pi 5; this seemed to alleviate high peak memory requirement. (I did increase shard counts for other generated files along the way, but I need to go back and figure out how much of that was strictly necessary vs. needing to use -j1 or -j2.)

Differential Revision: [D67925496](https://our.internmc.facebook.com/intern/diff/D67925496/)

[ghstack-poisoned]

swolchok commented:
[linux-jammy-py3-clang12-executorch](https://github.com/pytorch/pytorch/actions/runs/12699915549/job/35402147695?pr=144364#logs) job failed but is unstable and failing on main , so I'm ignoring the failure
